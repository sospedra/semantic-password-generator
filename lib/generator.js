'use strict';

var random = require('lodash.random');
var password = require('./password');
var transformNormalize = require('./transform/normalize');
var transformLength = require('./transform/length');
var transformLeet = require('./transform/leet');
var transformCase = require('./transform/case');
var transformRandom = require('./transform/random');
var transformSymbols = require('./transform/symbols');

var LENGTH_DEFAULT = 24;

module.exports = function (sentences) {
  var maxLength = password.getMaxLength(sentences);

  return function generator() {
    var inputLength = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : LENGTH_DEFAULT;

    var length = password.getLength(inputLength, maxLength);
    var data = sentences.filter(function (x) {
      return x.length >= length;
    });
    var base = data[random(data.length - 1)];

    if (!base) return null;

    return [transformNormalize(), transformLength(length), transformLeet(), transformCase(), transformRandom(), transformSymbols()].reduce(function (memo, fn) {
      return fn(memo);
    }, base);
  };
};