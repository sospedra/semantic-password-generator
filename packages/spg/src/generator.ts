import random from 'lodash.random'
import isNumber from 'lodash.isnumber'
import identity from 'lodash.identity'
import * as password from './password'
import transformLength from './transform/length'
import transformLeet from './transform/leet'
import transformCase from './transform/case'
import transformRandom from './transform/random'
import transformSymbols from './transform/symbols'

const DEFAULT_SENTENCES = [''] // allows init of all methods
const DEFAULT_ARGS = {
  case: false,
  length: 24,
  leet: false,
  random: false,
  symbols: false,
}
type Args = number | typeof DEFAULT_ARGS

const sanitizeOptions = (args: Args = {} as Args): typeof DEFAULT_ARGS => {
  return {
    ...DEFAULT_ARGS,
    ...(isNumber(args) ? { length: args } : args),
  }
}

export default function createGenerator(sentences = DEFAULT_SENTENCES) {
  const maxLength = password.getMaxLength(sentences)

  return function generator(args?: Args) {
    const options = sanitizeOptions(args)
    const length = password.getLength(options.length, maxLength)
    const data = sentences.filter((x) => x.length >= length)
    const base = data[random(data.length - 1)]

    if (!base) return null

    return [
      transformLength(length),
      options.leet && transformLeet(),
      options.case && transformCase(),
      options.random && transformRandom(),
      transformSymbols(options.symbols),
    ]
      .map((fn) => fn || identity)
      .reduce((memo, fn) => fn(memo), base)
  }
}
