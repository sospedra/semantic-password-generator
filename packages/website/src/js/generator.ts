import spg from 'semantic-password-generator'
import {
  $password,
  $renew,
  $slider,
  $hint,
  $case,
  $leet,
  $random,
  $symbols,
} from './elements'

const updateHint = (password: string) => {
  const length = password.length - 8

  if (length < 18) return ($hint.innerText = 'Weak')
  if (length < 24) return ($hint.innerText = 'Good')
  return ($hint.innerText = 'Strong ðŸ’ª')
}

export const addGenerator = async () => {
  let generator = await spg()

  const update = () => {
    const password = generator({
      length: Number($slider.value),
      case: $case.checked,
      leet: $leet.checked,
      random: $random.checked,
      symbols: $symbols.checked,
    })

    updateHint(password)
    $password.innerText = password
  }

  update()

  $renew.addEventListener('click', async () => {
    generator = await spg()
    update()
  })
  ;[$case, $leet, $random, $symbols, $slider].forEach(($) => {
    $.addEventListener('change', update)
  })
}
