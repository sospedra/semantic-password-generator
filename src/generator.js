const random = require('lodash.random')
const isNumber = require('lodash.isnumber')
const identity = require('lodash.identity')

const password = require('./password')
const transformNormalize = require('./transform/normalize')
const transformLength = require('./transform/length')
const transformLeet = require('./transform/leet')
const transformCase = require('./transform/case')
const transformRandom = require('./transform/random')
const transformSymbols = require('./transform/symbols')

const DEFAULT_SENTENCES = [0] // allows init of all methods

/**
 * Standarized options to set the base sentence and pipe transforms
 * @typedef {Object} Options
 * @property {Bool} case     [false]  - apply case transform
 * @property {Number} length [24]     - min and orientative sentence length
 * @property {Bool} leet     [false]  - apply leet transform
 * @property {Bool} random   [false]  - apply random transform
 * @property {Bool} symbols  [false]  - apply symbols transform
 */

/**
 * Transform incoming args into standarized options
 * @param  {Number|Object} args
 * @return {@type Options}
 */
const sanitizeOptions = (args) => {
  return Object.assign({}, {
    case: false,
    length: isNumber(args) ? args : 24,
    leet: false,
    random: false,
    symbols: false
  }, args)
}

module.exports = (sentences = DEFAULT_SENTENCES) => {
  const maxLength = password.getMaxLength(sentences)

  return function generator (args) {
    const options = sanitizeOptions(args)
    const length = password.getLength(options.length, maxLength)
    const data = sentences.filter((x) => x.length >= length)
    const base = data[random(data.length - 1)]

    if (!base) return null

    return [
      transformNormalize(),
      transformLength(length),
      options.leet && transformLeet(),
      options.case && transformCase(),
      options.random && transformRandom(),
      transformSymbols(options.symbols)
    ]
      .map((fn) => fn ? fn : identity)
      .reduce((memo, fn) => fn(memo), base)
  }
}
