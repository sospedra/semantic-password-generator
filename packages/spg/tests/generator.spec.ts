import generator from '../src/generator'
import stubs from './stubs'

test('generator is a function which returns null when no init', () => {
  expect(typeof generator).toBe('function')
  expect(generator()()).toBe(null)
})

test('generator transforms with default args when not provided', () => {
  stubs.forEach((sentence) => {
    expect(generator([sentence])()).toMatchSnapshot()
  })
})

test('generator takes one arg as number as length', () => {
  expect(generator([stubs.get('large')])(40)).toMatchSnapshot()
})
