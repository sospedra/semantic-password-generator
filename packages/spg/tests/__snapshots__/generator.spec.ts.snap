// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator takes one arg as number as length 1`] = `"this.is.my.long.sentence.to.try.and.to.check.the.length"`;

exports[`generator transforms with default args when not provided 1`] = `"this.is.my.long.sentence.to.try"`;

exports[`generator transforms with default args when not provided 2`] = `"a.bit.more.short.but.medium.long"`;

exports[`generator transforms with default args when not provided 3`] = `"super.short"`;

exports[`generator transforms with default args when not provided 4`] = `"itsn't.a.thing.we'll.understand"`;

exports[`generator transforms with default args when not provided 5`] = `"<p><b>Björk.Guðmundsdóttir</b>"`;

exports[`generator transforms with default args when not provided 6`] = `"<p><b>Rapid.Evolution</b>"`;

exports[`generator transforms with default args when not provided 7`] = `"Björk.Guðmundsdóttir.OTF.(/bjɜːrk/"`;
